### PR TITLE
Improved python tree plots

### PR DESCRIPTION
**Pull request by CharlesAuguste**
In this pull request, we tried to improve the way trees are plotted by the function  `plot_tree` . Visually digestible trees are essential for interpreting, debugging, and improving tree-based methods. LightGBM's current implementation of `plot_tree` produces trees which are very hard to mentally process and which, in our opinion, lack some useful information, like which nodes are constrained and how much data flows down through each node.

This PR makes several changes to the `plot_tree` function which increases the amount of information displayed in a given tree whilst, at the same time, making the tree more digestible.

Moreover, we add monotonically constrained features to the model's meta information so that this information is accessible through the model.

The figures below show how the output from the current `plot_tree` function compares with the output from our proposed `plot_tree` function, and, demonstrate how the proposed output contains significantly more information than the current output, yet, is much easier to interpret.

Here is a list of the new features from this pull request:
- Trees are now plotted horizontally and not vertically, and are more compact, which allows for a better visualization of big trees;
- Features and thresholds are now written in a more compact yet intelligible way;
- The default precision of the information printed is now 3;
- The user can now print the data percentage in nodes and leaves;
- Nodes splitted on a monotone feature are now colored in light green for a monotone increasing feature, and in light red for a monotone decreasing feature, and a legend is added to explain what these features correspond to;
 - Important information is printed in bold;
 - Other minor changes.

Current LightGBM tree:
image

Tree generated by the code of this pull request
image

### CODE REVIEW

**Comment by aldanor** [comment]
+@redditur

**Comment by aldanor** [comment]
@CharlesAuguste it looks like you have to update the tests (`tests/python_package_test/test_plotting.py`)

**Comment by aldanor** [comment]
Looks all green now :+1:

File: python-package/lightgbm/plotting.py [hunk 1]
```diff
@@ -380,44 +380,84 @@ def _to_graphviz(tree_info, show_info, feature_names, precision=None, **kwargs):
...
+            fillcolor = "white"
+            style = ""
+            if constraints:
+                if constraints[root['split_feature']] == 1:
+                    fillcolor = "#ddffdd"
+                    style = "filled" + style
+                if constraints[root['split_feature']] == -1:
+                    fillcolor = "#ffdddd"
+                    style = "filled" + style
+            label = "<" + label + ">"
```
[the remark this answers is not in the document]
Comment (CharlesAuguste) [reply python-package/lightgbm/plotting.py]: If I understand correctly, with graphiz, the label of nodes has to be HTML-Like, which means that it has to be an HTML string delimited by "<" and ">", as specified here https://graphviz.gitlab.io/_pages/doc/info/shapes.html#html. Some examples in the doc here https://graphviz.readthedocs.io/en/stable/manual.html also use it.

File: tests/python_package_test/test_plotting.py [hunk 2]
```diff
@@ -131,17 +131,27 @@ def test_create_tree_digraph(self):
...
+        self.assertIn('gain', graph_body)
+        self.assertIn('value', graph_body)
         self.assertIn('internal_weight', graph_body)
+        self.assertNotIn('leaf_data', graph_body)
         self.assertNotIn('internal_count', graph_body)
         self.assertNotIn('leaf_count', graph_body)
         self.assertNotIn('leaf_weight', graph_body)
 
+    @unittest.skipIf(not GRAPHVIZ_INSTALLED, 'graphviz is not installed')
+    def test_create_tree_digraph_with_monotone_constraints(self):
```
[the remark this answers is not in the document]
Comment (CharlesAuguste) [reply tests/python_package_test/test_plotting.py]: I can squash everything into a single test easily. However enabling monotone constraints for this dataset makes no sense (but still outputs something that can be plotted, so it is okay just for the sake of testing). So maybe the test would be a bit harder to understand, but I will do it if you think that is best.

[the remark this answers is not in the document]
Comment (redditur) [reply tests/python_package_test/test_plotting.py]: Yeah, I'm not convinced it make sense to bloat and convolute `tree_create_tree_digraph`. The test is asserting that the monotonic specific features work as expected, so it should remain separate.

File: python-package/lightgbm/plotting.py [hunk 3]
```diff
@@ -380,44 +380,84 @@ def _to_graphviz(tree_info, show_info, feature_names, precision=None, **kwargs):
     else:
         raise ImportError('You must install graphviz to plot tree.')
 
-    def add(root, parent=None, decision=None):
+    def add(root, total_count, parent=None, decision=None):
         """Recursively add node or edge."""
         if 'split_index' in root:  # non-leaf
+            l_dec, r_dec = 'is', "isn't"
```
[the remark this answers is not in the document]
Comment (redditur) [reply python-package/lightgbm/plotting.py]: +1 good idea.

File: tests/python_package_test/test_plotting.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply tests/python_package_test/test_plotting.py]: Test for monotone constraints differs from the original one only in adding monotone constraints (1 line) and asserting desired colors (2 lines). So I still feeling that it doesn't worth to add new test and waste CI time twice.

File: python-package/lightgbm/plotting.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 # pylint: disable = C0103
 """Plotting library."""
 from __future__ import absolute_import
+from __future__ import division
```
Comment (StrikerRUS) [inline python-package/lightgbm/plotting.py]: Please make it one-line.
```
from __future__ import absolute_import, division 
```

File: tests/python_package_test/test_plotting.py [hunk 5]
```diff
@@ -131,16 +132,14 @@ def test_create_tree_digraph(self):
...
-        self.assertIn('split_feature_name', graph_body)
-        self.assertNotIn('split_feature_index', graph_body)
-        self.assertIn('leaf_index', graph_body)
-        self.assertIn('split_gain', graph_body)
-        self.assertIn('internal_value', graph_body)
-        self.assertIn('internal_weight', graph_body)
-        self.assertNotIn('internal_count', graph_body)
-        self.assertNotIn('leaf_count', graph_body)
-        self.assertNotIn('leaf_weight', graph_body)
+        self.assertIn('leaf', graph_body)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_plotting.py]: Add
```
self.assertIn('feature', graph_body)
```
as well.

Comment (CharlesAuguste) [reply tests/python_package_test/test_plotting.py]: Unfortunately that wouldn't work as is. As you asked in a previous comment, I added the word "feature" for clarity, but just in the case where `feature_names` do not exist.
```
if feature_names is not None:
    label = '<B>{0}</B> {1} '.format(feature_names[root['split_feature']], operator)
else:
    label = 'feature <B>{0}</B> {1} '.format(root['split_feature'], operator)
```

In the case `feature_names` do exist it can be seen as superfluous either because the user specified the name itself so they should know what it is, either because default names are "Column_x", x being a number. In the test, names are not mentioned, but `feature_names` is not None (being and contains names of the form "Column_x"), so the word "feature" does not appear in the text. 

I can either add the word "feature" even when `feature_names` are not None, or forget the assertion. Let me know what is best.

Comment (StrikerRUS) [reply tests/python_package_test/test_plotting.py]: Sorry, my bad! Everything is OK with assertions.

**Comment by StrikerRUS** [comment]
I think it'll be better to make these lines consistent in terms of **boldness**. WDYT?

image

Another good visual change is swapping `weight` and `count` fields, because `count` and `data` have relationship, so it's better to have them nearby.


It's quite strange, but something is still wrong with error from monotone constraints.
```
import lightgbm as lgb
from sklearn.datasets import load_boston

X, y = load_boston(True)
lgb_train = lgb.Dataset(X, y, categorical_feature=[i for i in range(X.shape[1])])

params = {'mc': [1 for i in range(X.shape[1])]}
gbm = lgb.train(params, lgb_train)
```
Python package cannot catch raised error. Instead, kernel of Jupyter Notebook silently dies, and console logs seems to be weird:
```
[LightGBM] [Fatal] The output cannot be monotone with respect to categorical features
[LightGBM] [Fatal] The output cannot be monotone with respect to categorical feature[LightGBM] [Warning] s
The output cannot be monotone with respect to categorical features[LightGBM] [Fatal] The output cannot be monotone with re[LightGBM] [Warning] sThe output cannot be monotone with respect to categorical featurespect to categorical features

[LightGBM] [Fatal] The output cannot be monotone with respect to categorical features

[LightGBM] [Fatal] [LightGBM] [Warning] The output cannot be monotone with respect to categorical features
The output cannot be monotone with respect to categorical features[LightGBM] [Fatal] The output cannot be monotone with respect to categorical f
eatures
[LightGBM] [Warning] [LightGBM] [Fatal] The output cannot be monotone with respect to cateThe output cannot be monotone with respect to categorical featuresgorica
l[LightGBM] [Warning]  The output cannot be monotone with respect to categorical featuresfeatures
[LightGBM] [Fatal] The output cannot be monotone with respect to categorical features

[[LightGBM] [Warning] LThe output cannot be monotone with respect to categorical featuresi
ght[LightGBM] [Warning] GBM] [Fatal] The output cannot be monotone with respect to categorical featuresThe o
utpu[LightGBM] [Warning] t The output cannot be monotone with respect to categorical featuresca
nnot be monotone with respect to categorical features
[LightGBM] [Fatal] The output cannot be monotone with respect to categorical features
[LightGBM] [Warning] [LightGBM] [Fatal] [LightGBM] [Warning] The output cannot be monotone with respect to categorical features
The output cannot be monotone with respect to categorical features[LightGBM] [Fatal] [LightGBM] [Warning] The output cannot be monotone with respect to catThe output cannot be monotone with respect to categorical features

egorical features
The output cannot be monotone with respect to categorical features[[LightGBM] [Warning] LightGBM] [Fatal] The output cannot be monotone with respect to categorical features
The output cannot be monotone with respect to categorical features[LightGBM] [Warning]
The output cannot be monotone with respect to categorical features
```

I guess it's due to that this error is raised inside OMP loop. @guolinke Do we have any special treatment for that?

**Comment by CharlesAuguste** [comment]
Initially we only had features and leaf values in bold because we thought they were the most important information. But as you said, it is not consistent. In my opinion, a good compromise would be this (thresholds in bold added).
image

I don't think it makes sense to put "leaf x" in bold, because it is not an important information. But if you still prefer it that way I will make the necessary changes.

**Comment by CharlesAuguste** [comment]
Would you like this better ("Monotone constraints" added on top of the legend)?
image

**Comment by StrikerRUS** [comment]
@CharlesAuguste Nice, thanks! I think it looks clearer now.

**Comment by StrikerRUS** [comment]
@henry0312 @chivee @guolinke @jameslamb Can you please help and give a second review to this PR?

File: python-package/lightgbm/plotting.py [hunk 6]
```diff
@@ -380,48 +380,88 @@ def _to_graphviz(tree_info, show_info, feature_names, precision=None, **kwargs):
...
-            for info in show_info:
-                if info in {'split_gain', 'internal_value', 'internal_weight'}:
-                    label += r'\n{0}: {1}'.format(info, _float2str(root[info], precision))
-                elif info == 'internal_count':
-                    label += r'\n{0}: {1}'.format(info, root[info])
-            graph.node(name, label=label)
+            l_dec, r_dec = 'yes', "no"
             if root['decision_type'] == '<=':
-                l_dec, r_dec = '<=', '>'
+                operator = "&#8804;"
```
Comment (jameslamb) [inline python-package/lightgbm/plotting.py]: I think the code would be easier to understand if this was put in a named variable, e.g.

```python
lte_symbol = "&#8804;"
operator = lte_symbol
```

File: python-package/lightgbm/plotting.py [hunk 7]
```diff
@@ -380,48 +380,88 @@ def _to_graphviz(tree_info, show_info, feature_names, precision=None, **kwargs):
...
-            else:
-                label = 'split_feature_index: {0}'.format(root['split_feature'])
-            label += r'\nthreshold: {0}'.format(_float2str(root['threshold'], precision))
-            for info in show_info:
-                if info in {'split_gain', 'internal_value', 'internal_weight'}:
-                    label += r'\n{0}: {1}'.format(info, _float2str(root[info], precision))
-                elif info == 'internal_count':
-                    label += r'\n{0}: {1}'.format(info, root[info])
-            graph.node(name, label=label)
+            l_dec, r_dec = 'yes', "no"
```
Comment (jameslamb) [inline python-package/lightgbm/plotting.py]: in my opinion, simple assignments like this should be on their own lines. Could you change this to?

```python
l_dec = 'yes'
r_dec = 'no'
```

File: python-package/lightgbm/plotting.py [hunk 8]
```diff
@@ -441,8 +481,9 @@ def create_tree_digraph(booster, tree_index=0, show_info=None, precision=None,
     show_info : list of strings or None, optional (default=None)
         What information should be shown in nodes.
         Possible values of list items:
-        'split_gain', 'internal_value', 'internal_count', 'internal_weight', 'leaf_count', 'leaf_weight'.
-    precision : int or None, optional (default=None)
+        'split_gain', 'internal_value', 'internal_count', 'internal_weight',
+        'leaf_count', 'leaf_weight', 'data_percentage'.
+    precision : int or None, optional (default=3)
```
Comment (jameslamb) [inline python-package/lightgbm/plotting.py]: Could you expand this documentation please? What are the units and what is the implication of setting `None`?

File: python-package/lightgbm/plotting.py [hunk 9]
```diff
@@ -482,6 +523,11 @@ def create_tree_digraph(booster, tree_index=0, show_info=None, precision=None,
     else:
         feature_names = None
 
+    if 'monotone_constraints' in model:
+        monotone_constraints = model['monotone_constraints']
+    else:
+        monotone_constraints = None
```
Comment (jameslamb) [inline python-package/lightgbm/plotting.py]: could you please simplify this?

```python
monotone_constraints = model.get('monotone_constraints', None)
```

File: python-package/lightgbm/plotting.py (hunk 8 above)
Comment (StrikerRUS) [reply python-package/lightgbm/plotting.py]: @jameslamb This is nice catch, but I think this is beyond of this PR. `precision` was introduced in #1424 and is used in several functions, so I'll address your comment in a separate PR after merging this one.

Comment (jameslamb) [reply python-package/lightgbm/plotting.py]: makes sense!

**Comment by StrikerRUS** [comment]
Close-reopen to re-trigger Appveyor.

**Comment by StrikerRUS** [comment]
@jameslamb Can you please review the changes made?

**Comment by aldanor** [comment]
Great stuff, thanks guys 👍 (now we just need to figure out what to do with the other big PR to move it forward...)
